// zephyr-agent@0.0.31/dist/lib/http/upload-snapshot.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.uploadSnapshot = void 0;
const get_application_configuration_1 = require("../edge-requests/get-application-configuration");
const ze_http_request_1 = require("./ze-http-request");
const errors_1 = require("../errors");
const logging_1 = require("../logging");
async function uploadSnapshot({ body, application_uid, }) {
    const { EDGE_URL, jwt } = await (0, get_application_configuration_1.getApplicationConfiguration)({
        application_uid,
    });
    const json = JSON.stringify(body);
    const options = {
        method: 'POST',
        headers: {
            'Content-Length': Buffer.byteLength(json),
            'Content-Type': 'application/json; charset=utf-8',
            can_write_jwt: jwt,
        },
    };
    const [ok, cause, resp] = await ze_http_request_1.ZeHttpRequest.from({
        path: '/upload',
        base: EDGE_URL,
        query: {
            type: 'snapshot',
            skip_assets: true,
        },
    }, options, json);
    if (!ok) {
        throw new errors_1.ZephyrError(errors_1.ZeErrors.ERR_FAILED_UPLOAD, {
            type: 'snapshot',
            cause,
        });
    }
    (0, logging_1.ze_log)('Done: snapshot uploaded...', body);
    return resp;
}
exports.uploadSnapshot = uploadSnapshot;
