// zephyr-xpack-internal@0.0.31/dist/federation-dashboard-legacy/utils/convert-to-graph/convert-dependencies.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertDependencies = void 0;
function convertDependencies(params) {
    return {
        dependencies: convertDeps(params.npmModules, params.package.dependencies),
        devDependencies: convertDeps(params.npmModules, params.package.devDependencies),
        optionalDependencies: convertDeps(params.npmModules, params.package.optionalDependencies),
    };
}
exports.convertDependencies = convertDependencies;
function convertDeps(npmModules, deps) {
    if (!deps)
        return;
    return Object.entries(deps).map(([version, name]) => {
        const dataFromGraph = npmModules.get(name);
        const versionVal = version.replace(`${name}-`, '');
        if (dataFromGraph) {
            const foundInGraph = Object.values(dataFromGraph).find((depData) => depData.version.startsWith(versionVal));
            if (foundInGraph) {
                const { name, version, license, size } = foundInGraph;
                return { name, version, license, size };
            }
        }
        return {
            name,
            version: versionVal,
        };
    });
}
