// zephyr-agent@0.0.31/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.is_zephyr_dependency_pair = exports.ZephyrEngine = exports.zeBuildAssets = exports.zeBuildDashData = exports.buildAssetsMap = exports.logFn = exports.ze_log = exports.ZeErrors = exports.ZephyrError = exports.savePartialAssetMap = exports.removePartialAssetMap = exports.getPartialAssetMap = exports.resolveIndexHtml = exports.onIndexHtmlResolved = void 0;
// hack for angular
var resolve_index_html_1 = require("./lib/hacks/resolve-index-html");
Object.defineProperty(exports, "onIndexHtmlResolved", { enumerable: true, get: function () { return resolve_index_html_1.onIndexHtmlResolved; } });
Object.defineProperty(exports, "resolveIndexHtml", { enumerable: true, get: function () { return resolve_index_html_1.resolveIndexHtml; } });
// qwik hack - persist
var partial_assets_map_1 = require("./lib/node-persist/partial-assets-map");
Object.defineProperty(exports, "getPartialAssetMap", { enumerable: true, get: function () { return partial_assets_map_1.getPartialAssetMap; } });
Object.defineProperty(exports, "removePartialAssetMap", { enumerable: true, get: function () { return partial_assets_map_1.removePartialAssetMap; } });
Object.defineProperty(exports, "savePartialAssetMap", { enumerable: true, get: function () { return partial_assets_map_1.savePartialAssetMap; } });
// errors
var errors_1 = require("./lib/errors");
Object.defineProperty(exports, "ZephyrError", { enumerable: true, get: function () { return errors_1.ZephyrError; } });
Object.defineProperty(exports, "ZeErrors", { enumerable: true, get: function () { return errors_1.ZeErrors; } });
// logger
var logging_1 = require("./lib/logging");
Object.defineProperty(exports, "ze_log", { enumerable: true, get: function () { return logging_1.ze_log; } });
var ze_log_event_1 = require("./lib/logging/ze-log-event");
Object.defineProperty(exports, "logFn", { enumerable: true, get: function () { return ze_log_event_1.logFn; } });
// default transformers
var ze_build_assets_map_1 = require("./lib/transformers/ze-build-assets-map");
Object.defineProperty(exports, "buildAssetsMap", { enumerable: true, get: function () { return ze_build_assets_map_1.buildAssetsMap; } });
var ze_build_dash_data_1 = require("./lib/transformers/ze-build-dash-data");
Object.defineProperty(exports, "zeBuildDashData", { enumerable: true, get: function () { return ze_build_dash_data_1.zeBuildDashData; } });
var ze_build_assets_1 = require("./lib/transformers/ze-build-assets");
Object.defineProperty(exports, "zeBuildAssets", { enumerable: true, get: function () { return ze_build_assets_1.zeBuildAssets; } });
// Zephyr Edge is the main class which should be used
var zephyr_engine_1 = require("./zephyr-engine");
Object.defineProperty(exports, "ZephyrEngine", { enumerable: true, get: function () { return zephyr_engine_1.ZephyrEngine; } });
Object.defineProperty(exports, "is_zephyr_dependency_pair", { enumerable: true, get: function () { return zephyr_engine_1.is_zephyr_dependency_pair; } });
